kvserver: add `admission.io.overload` metric

KaiSun314 opened the pull request:
Resolves #87424.

Previously, only the unnormalized values of the LSM L0 sub-level and file counts is exposed externally, not the store's IOThreshold.

This was inadequate because it is tedious to normalize and compare the LSM L0 sub-level and file counts (as they require dividing by different numbers).

To address this, this patch adds a metric `admission.io.overload` tracking the store's IOThreshold.

Release note (ops change): Added a metric `admission.io.overload` which tracks the store's IOThreshold.

Note: I am not certain on the placement of the `admission.io.overload` metric, but it seems reasonable.
Justification:
1. Pretty much all updates to store metrics outside of ComputeMetrics in pkg/kv/kvstore/store.go seem to be increments (e.g. Inc(1) )
2. ComputeMetrics (there is a comment in the code: ComputeMetrics immediately computes the current value of store metrics which cannot be computed incrementally) updates the store metrics via updateCapacityGauges , updateReplicationGauges , updateEngineMetrics , and updateEnvStats. Looking at all these functions, updateReplicationGauges (where we are currently updating the IOOverload metric) seems to make the most sense to update the IOOverload metric
3. There seem to be two other admission metrics:
- RaftPausedFollowerCount - updated in updateReplicationGauges
- RaftPausedFollowerDroppedMsgs.Inc - updated in pkg/kv/kvserver/replica_raft.go via r.store.Metrics().RaftPausedFollowerDroppedMsgs.Inc(1)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87656)
<!-- Reviewable:end -->

kvoli reviewed (changes requested):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @KaiSun314 and @tbg)

___
*[`pkg/kv/kvserver/metrics.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NBlr3Uy4yUOJzTFb3zF:-NBlr3Uy4yUOJzTFb3zG:bo1w9zu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3b9ed4f9cb2d6bdefabf56ce486a94963a2d2aa/pkg/kv/kvserver/metrics.go#L1026)):*
> ```Go
> 	}
> 
> 	metaIOOverload = metric.Metadata{
> ```

We should also add it to the chart catalog - which should clear up some of the failing tests in CI.

Here:

https://github.com/cockroachdb/cockroach/blob/6309178fe9c85b57a27ff0c69d5adaeca3fafccb/pkg/ts/catalog/chart_catalog.go#L3505-L3500
___
*[`pkg/kv/kvserver/metrics.go` line 1028 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NBlrZtv2wAiEqX0XAtk:-NBlrZtv2wAiEqX0XAtl:b-vgntk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3b9ed4f9cb2d6bdefabf56ce486a94963a2d2aa/pkg/kv/kvserver/metrics.go#L1028)):*
> ```Go
> 	metaIOOverload = metric.Metadata{
> 		Name: "admission.io.overload",
> 		Help: `1-normalized float to pause replication to raft group followers if its value is at least 1.
> ```

nit: this feature isn't enabled by default - so to avoid any confusion we should qualify the statement.

See the cluster setting here:

https://github.com/cockroachdb/cockroach/blob/6309178fe9c85b57a27ff0c69d5adaeca3fafccb/pkg/kv/kvserver/replica_raft_overload.go#L27-L31


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @KaiSun314 and @tbg)

___
*[`pkg/kv/kvserver/store.go` line 3155 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NBls2bu9F6qWRSFQ4NR:-NBls2bu9F6qWRSFQ4NS:bl7qsy1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3b9ed4f9cb2d6bdefabf56ce486a94963a2d2aa/pkg/kv/kvserver/store.go#L3155)):*
> ```Go
> 	s.mu.RUnlock()
> 
> 	s.ioThreshold.Lock()
> ```

nit: let's add a 

```go
// TODO(kaisun314,kvoli): move this to a per-store admission control metrics struct when available. See pkg/util/admission/granter.go.
```

I don't think we should block the change on this but it would be good to unify them eventually.

See https://github.com/cockroachdb/cockroach/blob/6309178fe9c85b57a27ff0c69d5adaeca3fafccb/pkg/util/admission/granter.go#L710-L703 for more infor


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
Nice PR! A few comments. 

1. We should go over the failing essential CI and figure out what we are missing.
2. We should resolve the merge conflicts in `kv/kvserver/store.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-:-NBlsctm3eNBpnreRuck:bpbexy4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @KaiSun314 and @tbg)



<!-- Sent from Reviewable.io -->

KaiSun314 reviewed (commented):
Thank you for reviewing this PR!

1. The failing essential CIs now pass locally, will monitor to ensure they pass the required checks in Github
2. Merge conflicts should be resolved

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-:-NBnFDep4asHp2acvdnG:b-9hnhgw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/metrics.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NBlr3Uy4yUOJzTFb3zF:-NBmLtamEqBLYwGi1BOZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3b9ed4f9cb2d6bdefabf56ce486a94963a2d2aa/pkg/kv/kvserver/metrics.go#L1026)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

We should also add it to the chart catalog - which should clear up some of the failing tests in CI.

Here:

https://github.com/cockroachdb/cockroach/blob/6309178fe9c85b57a27ff0c69d5adaeca3fafccb/pkg/ts/catalog/chart_catalog.go#L3505-L3500
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

KaiSun314 committed:
kvserver: add admission.io.overload metric

Resolves #87424.

Previously, only the unnormalized values of the LSM L0 sub-level and
file counts is exposed externally, not the store's IOThreshold.

This was inadequate because it is tedious to normalize and compare the
LSM L0 sub-level and file counts (as they require dividing by different
numbers).

To address this, this patch adds a metric `admission.io.overload`
tracking the store's IOThreshold.

Release note (ops change): Added a metric `admission.io.overload` which
tracks the store's IOThreshold.

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-:-NC5We0v65XzkdLet54e:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @KaiSun314)



<!-- Sent from Reviewable.io -->

kvoli commented:
We should also backport this to 22.2.

let's chat on Monday about it. You can add the label on the rhs, backport-22.2

KaiSun314 commented:
> We should also backport this to 22.2.
> 
> let's chat on Monday about it. You can add the label on the rhs, backport-22.2

Sounds great! Added the backport-22.2.x label to this PR.

KaiSun314 commented:
bors r=kvoli

irfansharif reviewed (commented):
Nice work! This will come in handy in understanding AC behaviour (x-linking #82743). Just had a minor drive-by comment.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-:-NCa0yca2JfMTi4UhC-b:b-xn8h88)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/metrics.go` line 1033 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NCa-Kfl9W1xJBzNBziQ:-NCa-Kfl9W1xJBzNBziR:bom7bs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5571a275e44cd27727f390209c046dfb5d8518e/pkg/kv/kvserver/metrics.go#L1033)):*
> ```Go
> (pause replication feature is disabled if this setting is 0, feature is disabled by default);
> see pkg/kv/kvserver/replica_raft_overload.go for more details. Composed of LSM L0
> sub-level and file counts.`,
> ```

I found this help text a bit difficult to follow. Also since this is text that's available in any running binary (`SHOW ALL CLUSTER SETTINGS`), I wouldn't expect all readers to have the source code checked out, which would make the reference to this file slightly less than useful. It's also worth mentioning that even though we use this IO threshold for replication pausing, the metric is really it's own thing, so doesn't need to refer to it's one current use (there could be others in the future for ex., or that use could be removed but this metric kept around). I tried a following rephrasing:

```
1-normalized float indicating whether IO admission control considers the store as overloaded with respect to compaction out of L0.
```

In a code comment near pauseReplicationIOThreshold (admission.kv.pause_replication_io_threshold) we can mention that the metric it looks at underneath just so happens to be exported through admission.io.overload. If you end up trying this phrasing instead, do update the backport PR as well.
___
*[`pkg/kv/kvserver/metrics.go` line 1035 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87656#-NCa--dq0f-AtE-uSHDs:-NCa--dq0f-AtE-uSHDt:b-pz5j0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5571a275e44cd27727f390209c046dfb5d8518e/pkg/kv/kvserver/metrics.go#L1035)):*
> ```Go
> sub-level and file counts.`,
> 		Measurement: "Threshold",
> 		Unit:        metric.Unit_COUNT,
> ```

Should this be `metric.Unit_PERCENT` instead? That usually covers floats within [0.0, 1.0].


<!-- Sent from Reviewable.io -->
